analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs

```diff
@@ -236,21 +236,23 @@ private IEnumerable<ExplodedNode> ExceptionSuccessors(ExplodedNode node, Excepti
         state = node.Block.EnclosingRegion(ControlFlowRegionKind.Catch) is not null
             ? state.PushException(exception)        // If we're nested inside catch, we need to preserve the exception chain
             : state.SetException(exception);        // Otherwise we track only the current exception to avoid explosion of states after each statement
-        while (tryRegion is not null)
+
+        var handlersAreExhaustive = false;
+        while (tryRegion is not null && !handlersAreExhaustive)
         {
             var reachableHandlers = tryRegion.EnclosingRegion.NestedRegions.Where(x => x.Kind != ControlFlowRegionKind.Try && IsReachable(x, state.Exception)).ToArray();
             foreach (var handler in reachableHandlers)  // CatchRegion, FinallyRegion or FilterAndHandlerRegion
             {
+                handlersAreExhaustive |= CatchesAll(handler) || state.Exception.Type.DerivesFrom(handler.ExceptionType);
```

Comment:
Why was the style of implementation changed? The exhaustive condition now need to be checked twice, and this would be pretty easy to understand:
```
            if (handlersAreExhaustive)
            {
                yield break;
            }
```

Reply:
`yield break` caused us many headaches when understanding this method's code flow.

Reply:
Yes, it's a `return`. I often find `return` in loops to be hard to understand. And since it caused a lot of confusion for us in this case, I opted to rewrite it.